Source/Kernel/Grains/EventSequences/AppendedEventsQueue.cs

```diff
@@ -64,34 +69,87 @@ public Task Unsubscribe(ObserverKey observerKey)
     public void Dispose()
     {
         _queueTaskCompletionSource.SetCanceled();
+        _queueTask.Dispose();
+    }
+
+    /// <summary>
+    /// Await the queue to be depleted.
+    /// </summary>
+    /// <returns>Awaitable task.</returns>
+    /// <remarks>
+    /// This method will block until the queue is depleted. This is useful for testing purposes.
+    /// It is not exposed on the interface as it is not intended for production use.
+    /// </remarks>
+    public async Task AwaitQueueDepletion()
+    {
+        await Task.Run(async () =>
+        {
+            if (Debugger.IsAttached)
+            {
+                while (!_queue.IsEmpty)
+                {
+                    await Task.Delay(10);
+                }
+                await _queueEmptyEvent.WaitAsync();
+            }
+            else
+            {
+                var count = 10;
+                while (!_queue.IsEmpty)
+                {
+                    await Task.Delay(10);
+                    if (--count == 0)
+                    {
+                        break;
+                    }
+                }
+
+                await _queueEmptyEvent.WaitAsync().WaitAsync(TimeSpan.FromMilliseconds(500));
+            }
+        });
     }
 
     async Task QueueHandler()
     {
         while (!_queueTaskCompletionSource.Task.IsCanceled)
         {
             await _queueEvent.WaitAsync();
+            _queueEmptyEvent.Reset();
             if (_queueTaskCompletionSource.Task.IsCanceled)
             {
                 return;
             }
 
             while (_queue.TryDequeue(out var events))
             {
-                foreach (var subscription in _subscriptions)
+                try
                 {
-                    var actualEvents = events.Where(@event => subscription.EventTypeIds.Contains(@event.Metadata.Type.Id)).ToList();
-                    if (actualEvents.Count == 0)
+                    foreach (var group in events.GroupBy(@event => @event.Context.EventSourceId))
                     {
-                        continue;
+                        var tasks = new List<Task>();
+                        foreach (var subscription in _subscriptions)
+                        {
+                            var actualEvents = group.Where(@event => subscription.EventTypeIds.Contains(@event.Metadata.Type.Id)).ToList();
+                            if (actualEvents.Count == 0)
+                            {
+                                continue;
+                            }
+                            var observer = _grainFactory.GetGrain<IObserver>(subscription.ObserverKey);
+                            var partition = group.Key;
+                            tasks.Add(observer.Handle(partition, actualEvents));
+                        }
+
+                        await Task.WhenAll(tasks);
```

Comment:
While we're at it I think it could be useful to write an alternative path for when the events list only contains one event, which seems to be the case most of the time.

If we write a separate path for handling a single event we can avoid unnecessary allocations. It's a small performance optimization that in my opinion is pretty cheap :)